### DISCUSSION TITLE
Scope of bpftrace tools which are using 'kprobe' or 'uprobe'?

### DISCUSSION THREAD

**Discussion by rahul-kr4604** [bpftrace/bpftrace#1706] [Q&A] [answered]
In my deployment two containers 'test' (centos container) and 'test-bpf' (bpftrace container) are running inside the same pod 'test-pod'.

 **Observation:** I have observed that if any tool is using 'kprobe' then it captures all the events occur at host machine or inside container/pod running on same host machine and if any tool is using 'uprobe' then it captures only events which occur in the same user space where tool is executed i.e. only from host machine or only events from container/pod. 

Ex-
1. **'gethostlatency'** uses **uprobe** so it does not capture events from different user space. 
2. **'tcpconnect'** captures events from different user space because it uses **'kprobe'**. 

The tool 'gethostlatency' (which is executed inside the container 'test-bpf') does not capture any events from another container ('test' container) running inside the same pod. But when we execute 'tcpconnect' in the same environment then it captures the events from the other container running inside the pod and also captures the events outside the pod.

**Is my observation correct?**

**Comment by fbs** [comment, marked as answer]
Kernel probes trace the kernel. Tracing isn't namespaced so no matter what container you're in you'll still see all events.

Uprobes attach to a specific binary, relative to the current mount ns.

**Reply by rahul-kr4604** [reply]
Thanks @fbs for your reply!

I have some questions on this:

Two containers are running inside the same pod. Is it possible to capture events from one container to another container using a bpftrace tool which uses 'uprobe'?

If yes then, is it possible using mount namespace of containers?

Thanks....

**Reply by fbs** [reply]
With shared mount namespaces it might be possible, but I've never tried. I don't think the common runtimes support this kind of magic. 

I think the best option would be to run bpftrace in the root namespaces (on the host).  You can ofc still ship bpftrace in a container.
